zipkin-storage/elasticsearch-http/src/test/java/zipkin/storage/elasticsearch/http/ElasticsearchHttpSpanConsumerTest.java

```diff
@@ -266,6 +270,121 @@ public void addsPipelineId() throws Exception {
         .isEqualTo("/_bulk?pipeline=zipkin");
   }
 
+  @Test
+  public void indexesServiceSpanCache() throws Exception {
+    es.enqueue(new MockResponse());
+
+    Span span1 = TestObjects.TRACE.get(0).toBuilder().traceId(10L).build();
+    accept(span1);
+
+    assertThat(es.takeRequest().getBody().readByteString().utf8())
+      .contains(
+        "\"_type\":\"servicespan\",\"_id\":\"web|get\"}}\n"
+          + "{\"serviceName\":\"web\",\"spanName\":\"get\"}\n"
+      );
+
+    es.enqueue(new MockResponse());
+
+    Span span2 = TestObjects.TRACE.get(1).toBuilder().traceId(11L).build();
+    accept(span2);
+
+    assertThat(es.takeRequest().getBody().readByteString().utf8())
+      .contains(
+        "\"_type\":\"servicespan\",\"_id\":\"app|get\"}}\n"
+          + "{\"serviceName\":\"app\",\"spanName\":\"get\"}\n"
+      )
+      .doesNotContain(
+        "\"_type\":\"servicespan\",\"_id\":\"web|get\"}}\n"
+          + "{\"serviceName\":\"web\",\"spanName\":\"get\"}\n"
+      );
+  }
+
+  @Test
+  public void indexesServiceSpanCache_failedConcurrency() throws Exception {
+    final Integer threadCount = 10;
+    List<String> testElemsForCache = Arrays.asList("a", "b", "c", "d", "e", "f", "g", "h", "k", "l");
+    Boolean broken = false;
+    for (int j = 0; j < 10; j++) {
+      List<Thread> getOrSet = new ArrayList<>(threadCount * 2);
+      for (int i = 0; i < threadCount; i++) {
+        getOrSet.add(new Thread(
+          () -> testElemsForCache.forEach(elem ->
+            ElasticsearchHttpSpanConsumer.indexToServiceSpansCache.put(elem,
```

Comment:
I like the idea here, but I think you are testing map, not your code. If you encapsulate access to this indexToServiceSpansCache, you can test high-level methods you need (ex similar to how we have Multimap in InMemorySpanStore)